guide.md

```diff
@@ -0,0 +1,324 @@
+  nix_file_content = "(import <nixpkgs> {}).gcc11",
+  name = "nixpkgs_config_cc",
+)
+```
+
+`(import <nixpkgs> {})` loads the entire package set into an attribute set
+(Nix's equivalent of a record), and `.gcc11` selects one value from it. It is
+possible to customize the compiler further using more Nix if desired: you could,
+for instance, point Nix to a custom fork of Clang containing support for a new
+architecture, and have everything work seamlessly.
+
+With this, we now have a fully hermetic CC toolchain that will behave correctly
+across platforms. It remains to tell Bazel to use it by default, instead of
+`@local_config_cc`. To do so, we put the following into `.bazelrc` at the root
+of the repository:
+
+```
+build --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+run   --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+Now `bazel build` and `bazel run` should use this toolchain by default for
+C/C++ compilation:
+```
+> bazel build //src:hello-world
+INFO: Analyzed target //src:hello-world (42 packages loaded, 177 targets configured).
+INFO: Found 1 target...
+Target //src:hello-world up-to-date:
+  bazel-bin/src/hello-world
+INFO: Elapsed time: 0.377s, Critical Path: 0.22s
+INFO: 6 processes: 4 internal, 2 linux-sandbox.
+INFO: Build completed successfully, 6 total actions
+```
+
+# Making Nix optional
+
+It is entirely possible to make the use of Nix optional in a Bazel setup. To do
+so, we use `bazelrc` configs to only enable Nix-specific configuration when a
+`--config` flag is provided, like so:
+
+```
+build:nix --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+run:nix   --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build:nix --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+None of these options will apply to `bazel build` or `bazel run` unless an
+extra `--config nix` parameter is passed:
+
+```
+> bazel run //src:hello-world --config=nix
+...
+```
+
+# Working from a Nix-provided developer environment
+
+Another practical advantage of using Nix is that we can provide consistent,
+zero-setup development environments to all contributors to a project. These
+function similarly to, say, Python virtualenvs, but they're not just limited to
+providing dependencies for just one language ecosystem. Even better, they can
+provide system dependencies too.
+
+```
+> hello
+hello: command not found
+> nix-shell -p hello
+nix-shell> hello
+Hello, world!
+nix-shell> exit
+> hello
+hello: command not found
+```
+
+There's still a few loose ends in our setup that Nix developer environments can
+help tie up. First, the version of Bazel we've been using so far, has been
+unspecified. The general practice is to simply use one's distribution's package
+manager, something that doesn't deliver consistent Bazel versions across major
+OS platforms. Second, we haven't _disabled_ Bazel's deep-rooted tendency to 
```

Comment:
I think the more common practice outside of Nix is to use bazelisk, which can be configured to use a consistent Bazel version for the project.

Reply:
(Section taken out, see below.)